crates/sui-types/src/deny_list_v2.rs

```diff
@@ -28,58 +33,120 @@ struct ConfigKey {
+    fn get_type_tag() -> TypeTag {
+        TypeTag::Struct(Box::new(Self::type_()))
     }
 }
 
 /// Rust representation of the Move type 0x2::deny_list::AddressKey.
 #[derive(Debug, Serialize, Deserialize, Clone)]
 struct AddressKey(SuiAddress);
 
+impl AddressKey {
+    pub fn type_() -> StructTag {
+        StructTag {
+            address: SUI_FRAMEWORK_PACKAGE_ID.into(),
+            module: DENY_LIST_MODULE.to_owned(),
+            name: ident_str!("AddressKey").to_owned(),
+            type_params: vec![],
+        }
+    }
+}
+
 impl MoveTypeTagTrait for AddressKey {
     fn get_type_tag() -> TypeTag {
-        TypeTag::Struct(Box::new(StructTag {
+        TypeTag::Struct(Box::new(Self::type_()))
+    }
+}
+
+/// Rust representation of the Move type 0x2::deny_list::GlobalPauseKey.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct GlobalPauseKey;
+
+impl GlobalPauseKey {
+    pub fn type_() -> StructTag {
+        StructTag {
             address: SUI_FRAMEWORK_PACKAGE_ID.into(),
             module: DENY_LIST_MODULE.to_owned(),
-            name: ident_str!("AddressKey").to_owned(),
+            name: ident_str!("GlobalPauseKey").to_owned(),
             type_params: vec![],
-        }))
+        }
+    }
+}
+
+impl MoveTypeTagTrait for GlobalPauseKey {
+    fn get_type_tag() -> TypeTag {
+        TypeTag::Struct(Box::new(Self::type_()))
+    }
+}
+
+pub fn check_coin_deny_list_v2_during_signing(
+    address: SuiAddress,
+    input_objects: &CheckedInputObjects,
+    receiving_objects: &ReceivingObjects,
+    object_store: &dyn ObjectStore,
+) -> UserInputResult {
+    let coin_types = input_object_coin_types_for_denylist_check(input_objects, receiving_objects);
+    for coin_type in coin_types {
+        let Some(deny_list) = get_per_type_coin_deny_list_v2(&coin_type, object_store) else {
+            return Ok(());
+        };
+        if check_global_pause(&deny_list, object_store, None) {
+            return Err(UserInputError::CoinTypeGlobalPause { coin_type });
+        }
+        if check_address_denied_by_config(&deny_list, address, object_store, None) {
+            return Err(UserInputError::AddressDeniedForCoin { address, coin_type });
+        }
     }
+    Ok(())
 }
 
 pub fn get_per_type_coin_deny_list_v2(
-    coin_type: String,
+    coin_type: &String,
     object_store: &dyn ObjectStore,
 ) -> Option<Config> {
     let deny_list_root =
         get_deny_list_root_object(object_store).expect("Deny list root object not found");
-    let config: Config = get_dynamic_field_from_store(
-        object_store,
-        deny_list_root.id(),
-        &ConfigKey {
+    let config_key = DOFWrapper {
+        name: ConfigKey {
             per_type_index: DENY_LIST_COIN_TYPE_INDEX,
             per_type_key: coin_type.as_bytes().to_vec(),
         },
-    )
-    .ok()?;
+    };
+    let config: Config =
+        get_dynamic_field_from_store(object_store, deny_list_root.id(), &config_key).ok()?;
```

Comment:
this seems to deserialize the config object every time this is called. should we cache the deserialized value somewhere?